tests/transport/test_send_recv.cpp

```diff
@@ -0,0 +1,355 @@
+    if(rpeer_rank<0) rpeer_rank = size-1;
+
+    data_ptr = reinterpret_cast<int*>(smsg.data());
+    *data_ptr = rank;
+
+    timer.tic();
+    communicator_type::future<void> sreq, rreq;
+    for(int i=0; i<NITERS; i++){
+
+        rreq = comm.recv(rmsg, rpeer_rank, 1);
+        sreq = comm.send(smsg, speer_rank, 1);
+        while(!(rreq.ready() && sreq.ready()));
+
+        data_ptr = reinterpret_cast<int*>(rmsg.data());
+        EXPECT_TRUE(*data_ptr == rpeer_rank);
+    }
+    const auto t = timer.stoc();
+    if(rank==0)
+    {
+        std::cout << "time:       " << t/1000000 << "s\n";
+    }
+}
+
+template<typename MsgType, typename CommType>
+auto test_ring_send_recv_ft(CommType& comm) 
+{
+    MsgType smsg(4), rmsg(4);
+    test_ring_send_recv_ft(comm, std::move(rmsg), std::move(smsg));
+}
+
+TEST(transport, ring_send_recv_ft)
+{
+    auto context_ptr = gridtools::ghex::tl::context_factory<transport,threading>::create(1, MPI_COMM_WORLD);
+    auto& context = *context_ptr;
+    auto token = context.get_token();
+    auto comm = context.get_communicator(token);
+
+    test_ring_send_recv_ft<std::vector<unsigned char>>(comm);
+    test_ring_send_recv_ft<gridtools::ghex::tl::message_buffer<>>(comm);
+    test_ring_send_recv_ft<gridtools::ghex::tl::shared_message_buffer<>>(comm);
+    
+    {
+        gridtools::ghex::tl::message_buffer<> smsg(4), rmsg(4);
+        communicator_type::message_type asmsg{std::move(smsg)}, armsg{std::move(rmsg)};
+        test_ring_send_recv_ft(comm, std::move(asmsg), std::move(armsg));
+    }
+}
+
+
+
+template<typename MsgType, typename CommType>
+auto test_ring_send_recv_cb(CommType& comm, MsgType &&rmsg, MsgType &&smsg) 
+{
+    gridtools::ghex::timer timer;
+    int *data_ptr;
+    int rank = comm.rank();
+    int size = comm.size();
+    int speer_rank = (rank+1)%size;
+    int rpeer_rank = (rank-1)%size;
+    if(rpeer_rank<0) rpeer_rank = size-1;
+
+    data_ptr = reinterpret_cast<int*>(smsg.data());
+    *data_ptr = rank;
+
+    timer.tic();
+    volatile int received = 0;
+    volatile int sent = 0;
+    for(int i=0; i<NITERS; i++){
+
+        auto send_callback = [&](communicator_type::message_type, int, int) {sent++;};
+        auto recv_callback = [&](communicator_type::message_type, int, int) {received++;};    
+
+        comm.recv(rmsg, rpeer_rank, 1, recv_callback);
+        comm.send(smsg, speer_rank, 1, send_callback);
+        while(received<=i || sent<=i) comm.progress();
+
+        data_ptr = reinterpret_cast<int*>(rmsg.data());
+        EXPECT_TRUE(*data_ptr == rpeer_rank);
+    }
+
+    EXPECT_TRUE(received==NITERS && sent==NITERS);
+
+    const auto t = timer.stoc();
+    if(rank==0)
+    {
+        std::cout << "time:       " << t/1000000 << "s\n";
+    }
+}
+
+template<typename MsgType, typename CommType>
+auto test_ring_send_recv_cb(CommType& comm) 
+{
+    MsgType smsg(4), rmsg(4);
+    test_ring_send_recv_cb(comm, std::move(rmsg), std::move(smsg));
```

Comment:
@boeschf what will be the result of this and the `std::forward` from the previous suggestion?

Reply:
I think this is not what I want. In this cb test, as in the ft test, we do not give ownership of the message to ghex. It should operate on a message reference. My point here was to move the message into the actual `test_ring_send_recv_cb`, which runs the test. We can discuss this, because I guess I don't explain it well.

Reply:
If that's your intention, then you should not move. You can keep the `std::forward` above: it will not harm and it will work as expected if you at some point decided to transfer ownership to ghex. Personally, I think we should have both versions here, it is a test after all and the more cases we cover the better.